### ISSUE TITLE
MqttCommunicationTimedOutException

### ISSUE THREAD

**Issue by MattComb** [chkr1011/MQTTnet#167]
I have some test code which simply connects, subscribes and publishes a message. The subscription and publishing is done inside the connect event handler. Obviously this is not right but its just for the purpose of a test.

After approximately 10 executions of the connect code, I get the following timeout

"Exception of type 'MQTTnet.Exceptions.MqttCommunicationTimedOutException' was thrown."

I am just wondering if we can explain why this is occurring on such low volumes of connections with second delays between connection attempts.

```
       private async void Connect2()
        {
 
            var factory = new MqttFactory();
            var client = factory.CreateMqttClient();

            client.Connected += async (s, e) =>
            {
                // Subscribe to a topic
                await client.SubscribeAsync(new TopicFilterBuilder().WithTopic("topic123").Build());

                var payload = "{\"value\":\"" +GetRandomNumber(1,200).ToString()+"\"}";

                var message = new MqttApplicationMessageBuilder()
                    .WithTopic("topic123")
                    .WithPayload(payload)
                    .WithExactlyOnceQoS()
                    .WithRetainFlag()

                    .Build();

                await client.PublishAsync(message);

            };

            var options = new MqttClientOptionsBuilder()
                //.WithWebSocketServer("127.0.0.1:54251/mqtt")
                .WithWebSocketServer("localhost:1883/mqtt")
                //.WithTls()
                .Build();


            await client.ConnectAsync(options);
        }

        private void button2_Click(object sender, EventArgs e)
        {
            Connect2();
        }
```

**Comment by chkr1011** [comment]
Hi @MattComb,
Version 2.6.0 has a bug which slows down the connection rate so it can lead to timeouts etc.
Please test with version from develop or wait for 2.7.0 (either preview or final).

Best regards
Christian

**Comment by MattComb** [comment]
Thanks Christian, I'll give it a try!

**Comment by MattComb** [comment]
I have tried to progress this with the developer version of the code. I downloaded the dev branch and in order to rebuild I made the following two changes

1. I turned off code signing
2. I manually set the version number. (to 2.6.0 in order to work with the asp.net core application I had which was bound to that version)

On the client application I backed up the existing package and then replaced the platform targets with the folders from the MQTT framework bin output.

Currently, I am unable to connect at all, so I've done something wrong along the way, unfortunately.

Back to the original issue, I am a little concerned that the proposed solution for the connection issue (the delay in the finally area) might not be the only issue at play here as the timeout issue appears to be a degrade issue, it gets worse the more connections you open. e.g. If I connect 10 times, then even if I delay for 3 seconds before attempting another request it will timeout for me.

Is there any eta on the 2.7 nuget package to be released, I'm keen to test against a properly compiled package. a prerelease package in nuget would be perfect ?

Best Regards,

Matt.

**Comment by chkr1011** [comment]
Hi @MattComb,
the current _develop_ branch had some bugs which prevent reading packets at all. 
I will release a preview version tomorrow. Please let me know if this version works better.

Best regards
Christian

**Comment by MattComb** [comment]
Thats great news, thought I was going crazy. Ok I'll await the preview and I'll retry thanks.

**Comment by MattComb** [comment]
Hi Christian, I spent some time working on this today. I managed to get everything working correctly by wrapping the async code in a connection pool with a SemaphoreSlim.

For my testing I was parallelizing 20 connect->subscribe->publish tests. I think what was happening was when subsequent connections occurred, the previously connected clients were being dropped. (No they were not using shared variables, the code was thread safe). Does this sound like existing server functionality (i.e. one client session per clientoptions/client connection) ?

If this is correct, then I don't think I have raised anything new as the connection timeout is covered in another ticket.

I've now pooled the connections and it appears to be running well.

I will retest with the preview once its available.

Best Regards,

Matt.
